### ISSUE TITLE
Numerical Discrepancy Between Equivalent jnp.einsum Formulations

### ISSUE THREAD

**Issue by Liwink** [jax-ml/jax#29990]
### Description
I've observed a minor numerical difference when performing a mathematically equivalent operation using two different jnp.einsum patterns.

Method 1: Manually splitting a weight tensor and performing two separate `einsum` matrix multiplications.

Method 2: Using a single, combined `einsum` call that performs the same operations in a vectorized manner and unpacks the results.

I expected the outputs of both methods to be bit-for-bit identical. However, the results differ by one element out of ~7000 in my test case.

### Repro
https://colab.research.google.com/drive/1latj_SynZyqWxCKnwlhTegD1RMkRwpNP#scrollTo=P8yANYRKMwJw

### Discussion
I encounter this issue in my Jax implementation of Gemma 3 https://github.com/liwink/gemma-jax. The issue, which appears to stem from floating-point arithmetic, seems minor but significantly alters the final results. I need to determine if this is a bug or an expected consequence of floating-point behavior.

I am reporting this to confirm with the development team that this behavior is expected and to serve as a reference for others who might encounter a similar situation. Thank you!

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.2
jaxlib: 0.5.1
numpy:  2.0.2
python: 3.11.13 (main, Jun  4 2025, 08:57:29) [GCC 11.4.0]
device info: Tesla T4-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='3c9ee3b02273', release='6.1.123+', version='#1 SMP PREEMPT_DYNAMIC Sun Mar 30 16:01:29 UTC 2025', machine='x86_64')


$ nvidia-smi
Sat Jul  5 06:34:35 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   59C    P0             29W /   70W |   11442MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
```

**Comment by jakevdp** [comment]
At first glance, I suspect this is normal floating point behavior: when working with floating point math, it is expected that expressing the same mathematical operation in two different ways will not return bitwise equivalent results.

Your notebook checks for strict equality, and also uses `jnp.allclose`, which follows NumPy in having tolerances tuned to expected discrepancies in float64. It would not be surprising for these to return `False` if you were using, say `float32` or `bfloat16` operations.

If you think the differences are bigger than expected, I'd suggest putting together a full minimal repro so we can investigate: unfortunately the current version of your notebook provides no way for others to load the data you're using, no indication of what dtype you're using, and no way of seeing the magnitude of the discrepancies you're observing.

**Comment by Liwink** [comment]
Thank you Jake for your reply.

Yes, after double-checking the final result with softmax,  the differences are not bigger enough to change top 10 results. We can close this bug as WAI. Thanks again.
